### PR TITLE
Non empty shift

### PR DESCRIPTION
**Pull request by ChristopherDavenport**

### CODE REVIEW

File: modules/generic/src/main/scala/io/chrisdavenport/cormorant/generic/auto.scala [hunk 1]
```diff
@@ -8,7 +8,7 @@ import shapeless.labelled._
  * Fully Automatic Derivation of Any Product Type
   **/
 object auto {
-  implicit val hnilWrite = semiauto.hnilWrite
+  implicit def hnilWrite[H](implicit P: Put[H]): Write[H :: HNil] = semiauto.hnilWrite[H]
```
Comment (Daenyth) [inline modules/generic/src/main/scala/io/chrisdavenport/cormorant/generic/auto.scala]: Is it possible / sensible to add an ambiguous implicit for `HNil` by itself, with an implicitNotFound message pointing out the non-empty portion?

File: modules/parser/src/main/scala/io/chrisdavenport/cormorant/parser/package.scala [hunk 2]
```diff
@@ -21,8 +22,41 @@ package object parser {
...
+      if (headers.size > 1) {
+        CSV.Complete(h, filterLastRowIfEmpty(rows))
+      } else {
+        c
+      }
+    }
+  private def filterLastRowIfEmpty(rows: CSV.Rows): CSV.Rows = {
+    rows.rows.reverse match {
+      case x :: xl if x == CSV.Row(NonEmptyList(CSV.Field(""), Nil)) => CSV.Rows(xl.reverse)
+      case other => CSV.Rows(other.reverse)
```
Comment (Daenyth) [inline modules/parser/src/main/scala/io/chrisdavenport/cormorant/parser/package.scala]: `other.reverse` == `rows.rows` ? I don't think you need the extra reverse here.

File: modules/generic/src/main/scala/io/chrisdavenport/cormorant/generic/auto.scala (hunk 1 above)
Comment (ChristopherDavenport) [reply modules/generic/src/main/scala/io/chrisdavenport/cormorant/generic/auto.scala]: Due to the model of NonEmptyList, `HNil` by itself no longer makes much sense, instead we need to match on the last element that would be written and assert that this works and finishes off the HList.

I'm very open to better implicit not found instances.

Comment (Daenyth) [reply modules/generic/src/main/scala/io/chrisdavenport/cormorant/generic/auto.scala]: Right, the suggestion is to provide an *ambiguous* implicit for HNil to prompt scala to give the message. I suppose in retrospect it's not useful, as an empty type would simply fail to find one due to not having one for HNil

**Pull request merged**